pkg/apis/apps/validation/validation.go

```diff
@@ -153,12 +155,36 @@ func ValidateStatefulSetUpdate(statefulSet, oldStatefulSet *apps.StatefulSet) fi
 	restoreStrategy := statefulSet.Spec.UpdateStrategy
 	statefulSet.Spec.UpdateStrategy = oldStatefulSet.Spec.UpdateStrategy
 
+	restoreVolumeClaimTemplates := make([]api.PersistentVolumeClaim, len(statefulSet.Spec.VolumeClaimTemplates))
+	volumeExpansionEnabled := utilfeature.DefaultFeatureGate.Enabled(features.ExpandPersistentVolumes)
+	if len(oldStatefulSet.Spec.VolumeClaimTemplates) == len(statefulSet.Spec.VolumeClaimTemplates) {
+		for index, oldVolumeClaimTemplate := range oldStatefulSet.Spec.VolumeClaimTemplates {
+			restoreVolumeClaimTemplates[index] = *statefulSet.Spec.VolumeClaimTemplates[index].DeepCopy()
+			if volumeExpansionEnabled {
+				oldStorageRequest := oldVolumeClaimTemplate.Spec.Resources.Requests[api.ResourceStorage]
+				newStorageRequest := statefulSet.Spec.VolumeClaimTemplates[index].Spec.Resources.Requests[api.ResourceStorage]
+				if newStorageRequest.Cmp(oldStorageRequest) < 0 {
+					allErrs = append(allErrs, field.Forbidden(field.NewPath("spec", "volumeClaimTemplates", fmt.Sprint(index), "spec", "resources", "requests", "storage"), "storage request can not be less than previous value"))
+				}
+				statefulSet.Spec.VolumeClaimTemplates[index].Spec.Resources.Requests[api.ResourceStorage] = oldStorageRequest
+			}
+		}
+	}
+
 	if !apiequality.Semantic.DeepEqual(statefulSet.Spec, oldStatefulSet.Spec) {
-		allErrs = append(allErrs, field.Forbidden(field.NewPath("spec"), "updates to statefulset spec for fields other than 'replicas', 'template', and 'updateStrategy' are forbidden"))
+		var errMsg string
+		if volumeExpansionEnabled {
+			errMsg = "updates to statefulset spec for fields other than 'replicas', 'template', 'updateStrategy' and 'volumeClaimTemplate.Spec.Resources.Requests[storage]' are forbidden"
+		} else {
+			errMsg = "updates to statefulset spec for fields other than 'replicas', 'template' and 'updateStrategy' are forbidden"
```

Comment:
https://github.com/kubernetes/community/blob/master/contributors/devel/api_changes.md#backward-compatibility-gotchas:

> Validation rules on spec fields can neither be relaxed nor strengthened.

Reply:
This will make the entire thing unworkable isn't it? I think this must be brought to sig-arch and appropriately discussed. I am not sure I understand the reasoning of disallowing weakening of validations since we have a precedence of allowing such changes in past. For example: Volume resize in general would not have been possible if we followed argument to the letter.

Reply:
See the discussions here: https://github.com/kubernetes/kubernetes/issues/64841

> Weakening validation has the potential to break other consumers and generators of the API resource

Meanwhile @liggitt is working on concrete ways to tighten or loosen validation over multiple releases. Let's revisit this validation part later.